OpenRA.Mods.Common/Traits/Production/ProductionByDelivery.cs

```diff
@@ -0,0 +1,223 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System.Collections.Generic;
+using System.Drawing;
+using System.Linq;
+using OpenRA.Activities;
+using OpenRA.Mods.Common.Activities;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Deliver produced units from outside the map.")]
+	public class ProductionByDeliveryInfo : ProductionInfo
+	{
+		[Desc("Delivering actor name. The actor needs to have either of the `Helicopter` or `Plane` traits.")]
+		[ActorReference]
+		public readonly string DeliveryActor = "c17";
+
+		public readonly string ReadyAudio = "Reinforce";
+
+		public override object Create(ActorInitializer init) { return new ProductionByDelivery(init, this); }
+	}
+
+	class ProductionByDelivery : Production
+	{
+		readonly Actor self;
+		readonly CPos startPos;
+		readonly CPos endPos;
+		readonly ProductionByDeliveryInfo info;
+		readonly AircraftInfo aircraft;
+		readonly bool isPlane;
+
+		string factionVariant;
+
+		public ProductionByDelivery(ActorInitializer init, ProductionByDeliveryInfo info)
+			: base(init, info)
+		{
+			self = init.Self;
+
+			// Start a fixed distance away: the width of the map.
+			// This makes the production timing independent of spawnpoint
+			startPos = self.Location + new CVec(self.Owner.World.Map.Bounds.Width, 0);
+			endPos = new CPos(self.Owner.World.Map.Bounds.Left - 5, self.Location.Y);
+			this.info = info;
+
+			aircraft = self.World.Map.Rules.Actors[info.DeliveryActor].Traits.Get<AircraftInfo>();
+			isPlane = aircraft is PlaneInfo;
+		}
+
+		public override Actor DoProduction(Actor self, ActorInfo actorToProduce, ExitInfo exitInfo, string raceVariant)
```

Comment:
It would be really good if we didn't have to override the base implementation to make this simple again.  Could you extract the extra stuff that the base method does into the other outer DoProduction call so that this code can be shared?